### DISCUSSION TITLE
Performance vs Python high level API

### DISCUSSION THREAD

**Discussion by AlexandreBrown** [Kotlin/kotlindl#48] [General]
Hello,

In your opinion, let's say once KotlinDL 1.0+ is out, do you think there will be any difference (good or bad) performance-wise when compared to keras using Python high level api?
Another way to phrase it woud be: 
Since you mentioned KotlinDL uses low level Java APIs (JNI) to call native C++ directly, would it mean it would be faster than using Keras via Python? Is the overhead of Java lower than the overhead of Python? 
Or do you think there is no difference since Python does the same trick of letting the C++ do the heavy work ?

Thanks!

**Comment by zaleslaw** [comment]
1. Some thought about final performance: we benchmarked a few models. We found that it has the same speed as Keras models due to all heavy calculations in both cases are executed in TF Runtime written in C++, both Python and Kotlin call it.
2. About JNI: there is a very low JNI overhead (I measured it) due to rare calls (it takes 0.5-2% of the time), but we continue our experiments with alternative C/C++ wrapping like JavaCPP/maybe JEP 389: Foreign Linker API (Incubator). I hope it will be close to TF C++ API (maybe faster than pythonic Keras)
3. I have a little hope that we could be faster for the final user than Keras due to the ability to merge quickly new optimizers, SOTA models in Model Zoo, training tricks, and our own inference runtime based on ONNX (which is growing very fast)
